### PR TITLE
Constant-time square root

### PR DESCRIPTION
**Pull request by HastD**
This un-deprecates `Uint::sqrt` and provides a constant-time implementation. (Unfortunately this makes it non-`const`, so it's technically a breaking change; I don't see a way to handle zero without using `checked_div`, which is why this can't be `const`. For `const` contexts, `Uint::sqrt_vartime` can still be used.) The algorithm is just a slight modification of Algorithm 1.13 of Brent-Zimmermann as used in `sqrt_vartime`: instead of iterating until `xn` increases, we just iterate a fixed number of times (logarithmic in the number of bits) and then take the minimum of `guess` and `xn` at the end.

The reason this works is because of two properties of this algorithm that I prove here: [ct_sqrt.pdf](https://github.com/RustCrypto/crypto-bigint/files/12600669/ct_sqrt.pdf). (We can upload this somewhere more permanent—not sure where would be best. Also, it'd be great if someone could look this over to make sure I didn't make any mistakes in the proof.) First, we have an upper bound the number of iterations required. Second, once the algorithm reaches the desired value `sqrt(self)`, it will either repeat `sqrt(self)` or it will alternate between `sqrt(self)` and `sqrt(self) + 1`; thus, we can return the minimum of `guess` and `xn`.

I also removed some dead code in `Uint::sqrt_vartime`: unless I'm really missing something here, the first `while` loop can never run because the initial value of `guess` is defined such that it's never an underestimate.

### CODE REVIEW

**Comment by tarcieri** [comment]
> We can upload this somewhere more permanent—not sure where would be best.

arXiv perhaps?

**Comment by HastD** [comment]
Sure, I can upload it to arXiv (cs.DS probably?). I'd like to wait until at least one other person has looked it over it reduce the chance of errors slipping through, but after that I'll go ahead.

**Comment by tarcieri** [comment]
cc @fjarri

File: src/uint/sqrt.rs [hunk 1]
```diff
@@ -1,17 +1,36 @@
...
+    ///
+    /// Callers can check if `self` is a square by squaring the result
+    pub fn sqrt(&self) -> Self {
+        let max_bits = (self.bits() + 1) >> 1;
+        let cap = Self::ONE.shl(max_bits);
+        let mut guess = cap; // ≥ √(`self`)
+        let mut xn = {
+            let q = self.wrapping_div(&guess);
+            let t = guess.wrapping_add(&q);
+            t.shr_vartime(1)
```
Comment (fjarri) [inline src/uint/sqrt.rs:18]: Should be just `shr()`

**Comment by fjarri** [comment]
> I don't see a way to handle zero without using `checked_div`, which is why this can't be const

I am not sure I see the problem here, you can always use the underlying `Uint::ct_div_rem()` and use the returned `CtChoice` to conditionally select the result.

File: src/uint/sqrt.rs (hunk 1 above)
Comment (fjarri) [reply src/uint/sqrt.rs:18]: Actually no, disregard that. My mistake.

**Comment by HastD** [comment]
> I am not sure I see the problem here, you can always use the underlying `Uint::ct_div_rem()` and use the returned `CtChoice` to conditionally select the result.

I just overlooked that, so there's no obstruction to keeping it `const`. Actually, it looks like `Uint::ct_div_rem` [already internally does that conditional selection](https://github.com/RustCrypto/crypto-bigint/blob/master/src/uint/div.rs#L75) (so the returned quotient is zero if `rhs` is zero); can that behavior be relied on or is it an internal implementation detail of `ct_div_rem`? For reference, the relevant lines of code in `ct_div_rem`:

```rust
    pub(crate) const fn ct_div_rem(&self, rhs: &Self) -> (Self, Self, CtChoice) {
        let mb = rhs.bits_vartime();
        // [...]
        let is_some = Limb(mb as Word).ct_is_nonzero();
        quo = Self::ct_select(&Self::ZERO, &quo, is_some);
        (quo, rem, is_some)
    }
```

File: src/uint/sqrt.rs [hunk 2]
```diff
@@ -1,17 +1,36 @@
...
+    /// Computes √(`self`) in constant time.
+    /// Based on Brent & Zimmermann, Modern Computer Arithmetic, v0.5.9, Algorithm 1.13
+    ///
+    /// Callers can check if `self` is a square by squaring the result
+    pub fn sqrt(&self) -> Self {
+        let max_bits = (self.bits() + 1) >> 1;
+        let cap = Self::ONE.shl(max_bits);
+        let mut guess = cap; // ≥ √(`self`)
+        let mut xn = {
+            let q = self.wrapping_div(&guess);
```
[the remark this answers is not in the document]
Comment (HastD) [reply src/uint/sqrt.rs:16]: My mistake, I didn't see that `wrapping_div` is only constant-time for a fixed `rhs`. Actually, I think there might be a documentation issue here—I can't find anywhere in the public documentation that says this, only the documentation for `ct_div_rem` (which doesn't appear in the public docs since it's `pub(crate)`).

What do you think would be the better approach here: making a new function that's like `ct_div_rem` but constant-time with respect to both inputs, or modifying `ct_div_rem` to have this stronger constant-time guarantee and moving the "constant-time only for fixed rhs" behavior to a new function?

[the remark this answers is not in the document]
Comment (fjarri) [reply src/uint/sqrt.rs:16]: I agree that the name is misleading - it should have had a `vartime` suffix (and the whole `div.rs` is kind of a mess in terms of naming - see #268). So the proper way to proceed I think would be to rename the current one to `_vartime`, and implement a constant-time one in its place.

**Comment by fjarri** [comment]
> can that behavior be relied on or is it an internal implementation detail of `ct_div_rem`

I think it should be treated as an implementation detail.

**Comment by fjarri** [comment]
Also:
- I'll go through your proof in a few days, looks like it'll require some time and concentration :)
- Perhaps a proptest (see `tests/proptests.rs`) would be good to add here, for both vartime and const-time versions.

**Comment by HastD** [comment]
Okay, I just made the necessary changes so `sqrt` and `wrapping_sqrt` are `const` functions again, and I modified the `wrapping_sqrt` proptest to test both the vartime and constant-time versions. However, as discussed above, it's still not actually constant-time yet because `ct_div_rem` (and hence also `wrapping_div`) are currently only constant-time for fixed divisor.

**Comment by HastD** [comment]
Okay, I added an implementation of division that's constant-time in both arguments, and moved the old division implementation (that's constant-time only for fixed divisor) to `_vartime` functions (`div_rem_vartime`, `wrapping_div_vartime`, etc). If this has been implemented correctly then `Uint::sqrt` should be fully constant-time now.

My constant-time division implementation is a straightforward modification of the vartime algorithm, just replacing vartime methods with their constant-time variants, and running for the maximum number of iterations with a `ct_select` to avoid modifying the outputs beyond when the loop would've terminated in the vartime version. I suspect it's possible to do this more efficiently but I'll put this out there as a starting point. How's it look?

File: src/uint/sqrt.rs [hunk 3]
```diff
@@ -1,17 +1,39 @@
...
+            let t = guess.wrapping_add(&q);
+            t.shr_vartime(1)
+        };
+
+        // Repeat enough times to guarantee result has stabilized.
+        // See Hast, "Note on computation of integer square roots" for a proof of this bound.
+        let mut i = 0;
+        while i < usize::BITS - Self::BITS.leading_zeros() {
```
Comment (fjarri) [inline src/uint/sqrt.rs:24]: Self::LOG2_BITS could be used here

File: src/uint/sqrt.rs [hunk 4]
```diff
@@ -1,17 +1,39 @@
...
+        let cap = Self::ONE.shl(max_bits);
+        let mut guess = cap; // ≥ √(`self`)
+        let mut xn = {
+            let q = self.wrapping_div(&guess);
+            let t = guess.wrapping_add(&q);
+            t.shr_vartime(1)
+        };
+
+        // Repeat enough times to guarantee result has stabilized.
+        // See Hast, "Note on computation of integer square roots" for a proof of this bound.
```
Comment (fjarri) [inline src/uint/sqrt.rs:22]: When your paper has a more permanent link (e.g. on arxiv), please make a PR referencing it here

File: src/uint/sqrt.rs [hunk 5]
```diff
@@ -1,17 +1,39 @@
...
+        let mut i = 0;
+        while i < usize::BITS - Self::BITS.leading_zeros() {
+            guess = xn;
+            xn = {
+                let (q, _, is_some) = self.const_div_rem(&guess);
+                let q = Self::ct_select(&Self::ZERO, &q, is_some);
```
Comment (fjarri) [inline src/uint/sqrt.rs:28]: Is this needed specifically to handle the case of `self == 0`?

File: src/uint/sqrt.rs [hunk 6]
```diff
@@ -1,17 +1,39 @@
...
+                let t = guess.wrapping_add(&q);
+                t.shr_vartime(1)
+            };
+            i += 1;
+        }
+
+        // at least one of `guess` and `xn` is now equal to √(`self`), so return the minimum
```
Comment (fjarri) [inline src/uint/sqrt.rs:35]: So at this point `guess == x_n`, and `xn == x_{n+1}`, where `n = floor(log2(Self::BITS))`. But in the paper it says that it should be `n = floor(log2(Self::BITS)) + 1` - am I missing something?

File: src/uint/sqrt.rs [hunk 7]
```diff
@@ -23,31 +45,17 @@ impl<const LIMBS: usize> Uint<LIMBS> {
...
-            xn = {
-                let q = self.wrapping_div(&guess);
-                let t = guess.wrapping_add(&q);
-                t.shr_vartime(1)
-            };
-        }
+        // Note, xn <= guess at this point.
 
         // Repeat while guess decreases.
         while Uint::ct_gt(&guess, &xn).is_true_vartime() && xn.ct_is_nonzero().is_true_vartime() {
```
Comment (fjarri) [inline src/uint/sqrt.rs:55]: We don't need to use `ct_gt()` and `ct_is_nonzero()` here, those are constant-time

File: src/uint/sqrt.rs [hunk 8]
```diff
@@ -56,29 +64,26 @@ impl<const LIMBS: usize> Uint<LIMBS> {
         Self::ct_select(&Self::ZERO, &guess, self.ct_is_nonzero())
```
Comment (fjarri) [inline src/uint/sqrt.rs:64]: Similarly here, we don't need constant-timeness.

**Comment by fjarri** [comment]
Sorry for the delay, I was kind of swamped at work. I've read the paper, and it looks sound to me (although a bit too terse :). Have you submitted it anywhere yet? Would be nice to have a reference in the comments to the code. 

I've also got some relatively minor comments to the code.

**Comment by tarcieri** [comment]
> I've read the paper, and it looks sound to me (although a bit too terse :). Have you submitted it anywhere yet? Would be nice to have a reference in the comments to the code.

I think with a bit more polish on it you could throw it on the IACR ePrint service.

File: src/uint/sqrt.rs [hunk 9]
```diff
@@ -103,13 +108,24 @@ mod tests {
 
     #[test]
     fn edge() {
+        assert_eq!(U256::ZERO.sqrt(), U256::ZERO);
+        assert_eq!(U256::ONE.sqrt(), U256::ONE);
+        let mut half = U256::ZERO;
+        for i in 0..half.limbs.len() / 2 {
+            half.limbs[i] = Limb::MAX;
+        }
+        assert_eq!(U256::MAX.sqrt(), half);
```
Comment (fjarri) [inline src/uint/sqrt.rs:117]: An idea for the edge case test: a number that actually needs the maximum amount of iterations to converge. According to my tests, the ones before 10000 are 80, 99, 4224, 4355, 4488, 4623, 4760, 4899; but please check independently. (Also an interesting mathematical question - is there some rule for their distribution)

File: src/uint/sqrt.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tarcieri) [reply src/uint/sqrt.rs:16]: It appears this is a blocker on merging this PR. I guess we can go ahead and flip over to the v0.6 series per #268 and try to land this PR afterward.

[the remark this answers is not in the document]
Comment (tarcieri) [reply src/uint/sqrt.rs:16]: Checking on this: `wrapping_div` calls `const_div_rem`, which claims:

```rust
    /// This function is constant-time with respect to both `self` and `rhs`.
    pub(crate) const fn const_div_rem(&self, rhs: &Self) -> (Self, Self, CtChoice) {
```

[the remark this answers is not in the document]
Comment (fjarri) [reply src/uint/sqrt.rs:16]: Which commit are you looking at? The current `master` has no `const_div_rem()`, and `Uint::wrapping_div()` uses `ct_div_rem()`, which is not constant-time in `rhs`.

[the remark this answers is not in the document]
Comment (tarcieri) [reply src/uint/sqrt.rs:16]: This PR changes `wrapping_div` to call `const_div_rem`, which is also added by this PR

**Comment by tarcieri** [comment]
Merged in #376
